utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+  .   fst
+  <*> fromMaybe []
+  .   ubos
+  .   snd
+  <*> fromMaybe []
+  .   textures
+  .   snd
+  )
+
+makeInputInfo
+  :: Vector (Shader, Reflection)
+  -> Maybe (SomeStruct PipelineVertexInputStateCreateInfo)
+makeInputInfo =
+  (SomeStruct <$>)
+    . makePipelineVertexInputStateCreateInfo
+    . join
+    . V.fromList
+    . catMaybes
+    . (inputs . snd <$>)
+    . filter ((== Vert) . (stage :: Shader -> ShaderStage) . fst)
+    . V.toList
+
+makeShaderModuleCreateInfo
+  :: "code" ::: B.ByteString -> ShaderModuleCreateInfo '[]
+makeShaderModuleCreateInfo code = zero { code = code }
+
+makePipelineShaderStageCreateInfo
+  :: ShaderStage
+  -> ShaderModule
+  -> EntryPoint
+  -> PipelineShaderStageCreateInfo '[]
+makePipelineShaderStageCreateInfo stage shaderModule EntryPoint {..} = zero
+  { stage   = shaderStageFlagBits stage
+  , module' = shaderModule
+  , name    = B.pack . T.unpack $ name
+  }
+
+makePipelineShaderStageCreateInfos
+  :: ShaderStage
+  -> Vector EntryPoint
+  -> ShaderModule
+  -> Vector (PipelineShaderStageCreateInfo '[])
+makePipelineShaderStageCreateInfos stage entryPoints shaderModule =
+  V.map (makePipelineShaderStageCreateInfo stage shaderModule) entryPoints
+
+-- https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt
+-- https://vulkan.lunarg.com/doc/view/1.2.135.0/linux/tutorial/html/08-init_pipeline_layout.html
+makeUboDescriptorSetLayoutBinding
+  :: ShaderStage -> Ubo -> (Int, DescriptorSetLayoutBinding)
+makeUboDescriptorSetLayoutBinding stage Ubo {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = DESCRIPTOR_TYPE_UNIFORM_BUFFER
+         , descriptorCount = maybe 1 (V.sum . (fromIntegral <$>)) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+data TextureDescriptorType = Sampler2D
+  deriving Show
+
+textureDescriptorTypeFromText :: Text -> Maybe TextureDescriptorType
+textureDescriptorTypeFromText = \case
+  "sampler2D" -> Just Sampler2D
+  _           -> Nothing
+
+textureDescriptorTypeToText :: TextureDescriptorType -> Text
+textureDescriptorTypeToText = \case
+  Sampler2D -> "sampler2D"
+
+instance FromJSON TextureDescriptorType where
+  parseJSON = withTextMaybe "type" textureDescriptorTypeFromText
+
+instance ToJSON TextureDescriptorType where
+  toJSON     = String . textureDescriptorTypeToText
+  toEncoding = toEncoding . textureDescriptorTypeToText
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding
+  :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = convertTextureDescriptorType type'
+         , descriptorCount = maybe 1 (V.sum . (fromIntegral <$>)) array
```

Comment:
If I understand this correctly, you can `fromIntegral` only once - right before setting `descriptorCount`.

Reply:
fixed, now it is `maybe 1 (fromIntegral . V.sum) array`